
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://pgcbioinfo.github.io/vargram/basics/">
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../mutation_profile/">
      
      
      <link rel="icon" href="../assets/images/transparent_colored.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Quickstart - VARGRAM</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="custom">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#quickstart" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="VARGRAM" class="md-header__button md-logo" aria-label="VARGRAM" data-md-component="logo">
      
  <img src="../assets/images/transparent_colored.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            VARGRAM
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Quickstart
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="custom"  aria-label="Dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="black" data-md-color-accent="custom"  aria-label="Light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/pgcbioinfo/vargram" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
  
  About

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  
  Tutorial

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../install_vargram/" class="md-tabs__link">
          
  
  
  Installation

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      
  
  
  
  
    
    
      
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../reference/vargram/data/example/" class="md-tabs__link">
          
  
  
  Code Reference

        </a>
      </li>
    
  

    
  

    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="VARGRAM" class="md-nav__button md-logo" aria-label="VARGRAM" data-md-component="logo">
      
  <img src="../assets/images/transparent_colored.png" alt="logo">

    </a>
    VARGRAM
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/pgcbioinfo/vargram" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    About
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            About
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    VARGRAM
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Tutorial
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Tutorial
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Quickstart
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Quickstart
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-syntax" class="md-nav__link">
    <span class="md-ellipsis">
      Basic syntax
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#vargram-methods" class="md-nav__link">
    <span class="md-ellipsis">
      VARGRAM methods
    </span>
  </a>
  
    <nav class="md-nav" aria-label="VARGRAM methods">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#types" class="md-nav__link">
    <span class="md-ellipsis">
      Types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#order" class="md-nav__link">
    <span class="md-ellipsis">
      Order
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../mutation_profile/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Generating a mutation profile
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Installation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../install_vargram/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    VARGRAM
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../install_nextclade/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Nextclade CLI
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Code Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    vargram
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            vargram
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    data
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/data/example/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    example
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/data/nextclade/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    nextclade
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    plots
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            plots
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/plots/_profile/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    _profile
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/plots/_profile_elements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    _profile_elements
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/plots/_profile_renderer/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    _profile_renderer
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/vargram/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    vargram
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_4" >
        
          
          <label class="md-nav__link" for="__nav_4_1_4" id="__nav_4_1_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    wranglers
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_4">
            <span class="md-nav__icon md-icon"></span>
            wranglers
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/wranglers/_nextclade/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    _nextclade
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/wranglers/_nextclade_cli/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    _nextclade_cli
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/wranglers/_nextclade_utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    _nextclade_utils
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/vargram/wranglers/_wrangler/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    _wrangler
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-syntax" class="md-nav__link">
    <span class="md-ellipsis">
      Basic syntax
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#vargram-methods" class="md-nav__link">
    <span class="md-ellipsis">
      VARGRAM methods
    </span>
  </a>
  
    <nav class="md-nav" aria-label="VARGRAM methods">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#types" class="md-nav__link">
    <span class="md-ellipsis">
      Types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#order" class="md-nav__link">
    <span class="md-ellipsis">
      Order
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="quickstart">Quickstart</h1>
<h2 id="example">Example</h2>
<p>If you wish to get started immediately, <a href="../install_vargram/">install VARGRAM</a> (<code>pip install vargram</code>). Then, run either of the following code snippets on a Jupyter notebook or as a Python script. The code below downloads <a href="https://github.com/pgcbioinfo/vargram/tree/main/tests/test_data">test data from the VARGRAM repo</a> and does not require <a href="https://pgcbioinfo.github.io/vargram/install_nextclade/">Nextclade CLI</a>.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Simple script for bare figure</label><label for="__tabbed_1_2">Modified script for more customized figure</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p>The following is the simplest way to generate a mutation profile with default settings:
<div class="highlight"><pre><span></span><code><span class="c1"># Import main VARGRAM module and module to download external data</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">vargram</span><span class="w"> </span><span class="kn">import</span> <span class="n">vargram</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">vargram.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">example</span>

<span class="c1"># Download test data into directory called test_data </span>
<span class="c1"># Directory is created if it does not yet exist</span>
<span class="n">example</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;test_data&#39;</span><span class="p">)</span> 

<span class="c1"># Generate the mutation profile</span>
<span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;test_data/analysis/omicron_analysis_cli.tsv&#39;</span><span class="p">)</span> <span class="c1"># Provide data</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span> <span class="c1"># Tell VARGRAM you want to create a mutation profile</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> <span class="c1"># Show the figure</span>
<span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;default_profile.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span> <span class="c1"># Save the figure</span>
</code></pre></div>
This will produce the following bare figure:
<img alt="plot" src="../assets/images/default_profile.png" /></p>
<p>The profile consists of several barplots stacked together; there is one barplot per mutated gene (gene name is shown in rectangular box just above the barplot). The x-axis of each barplot shows the amino acid mutations (substitutions, deletions, and insertions) in the corresponding gene and the y-axis shows either the raw number ("counts") or the proportion ("weights") of sequences with those mutations. The bars are colored per sequence batch.</p>
<p>By default, the batch labels in the legend are the filenames of the provided FASTA files or <code>my_batch</code> if a raw Nextclade analysis CSV file is provided instead. In the above example, the file <code>omicron_analysis_cli.tsv</code> is a concatenation of two Nextclade analysis files with an added <code>batch</code> column, which VARGRAM reads (see <a href="mutation_profile/#other-features">tutorial for specifying batch column and other features</a>). Some gene names may be too long to fit within the label above its corresponding bar plot. In which case, the gene name is replaced with a number and a legend is automatically created for it.</p>
<details class="question">
<summary>How are the weights calculated?</summary>
<p>When multiple batches are detected, the y-axis shows the weighted count of a mutation by default. Per batch, the weight of a mutation is defined as its count (no. of occurences) in the batch divided by all mutation counts in the same batch. Thus, a mutation E156G having a weight of 3 in a batch means that 3% of all mutation counts in that batch are E156G mutations. </p>
<p>This choice was made so that two batches with different number of samples can still be compared. The y-axis can be changed to show raw counts even if there are multiple batches.</p>
</details>
<details class="question">
<summary>Why are the genes not ordered?</summary>
<p>By default, the genes are shown based roughly on the no. of mutations it has and these genes are then packed to maintain a rectangular aspect ratio. To get a profile with ordered genes, a genome annotation file must be provided. A horizontal layout of the barplots can also be specified.</p>
</details>
</div>
<div class="tabbed-block">
<p>To get a more interesting figure, run the following instead which modifies the labels and the colors, among other things, and adds key mutation heatmaps for reference:
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">vargram.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">example</span> <span class="c1"># Import module for downloading external data</span>
<span class="n">example</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;test_data&#39;</span><span class="p">)</span> <span class="c1"># Download test data into directory called test_data </span>

<span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;test_data/analysis/omicron_analysis_cli.tsv&#39;</span><span class="p">)</span> <span class="c1"># Provide data</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="c1"># Set minimum count for a mutation to be included</span>
        <span class="n">ytype</span><span class="o">=</span><span class="s1">&#39;counts&#39;</span><span class="p">)</span> <span class="c1"># Set y-axis to show raw count</span>
<span class="n">vg</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">stack_title</span><span class="o">=</span><span class="s1">&#39;Region&#39;</span><span class="p">,</span> <span class="c1"># Change batch legend title</span>
<span class="n">stack_label</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Foreign&#39;</span><span class="p">,</span> <span class="s1">&#39;Local&#39;</span><span class="p">],</span> <span class="c1"># Change batch names</span>
<span class="n">stack_color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#009193&#39;</span><span class="p">,</span> <span class="s1">&#39;#E33E84&#39;</span><span class="p">])</span> <span class="c1"># Change batch bar colors</span>
<span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;test_data/keys/BA1_key.csv&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;BA.1&#39;</span><span class="p">)</span> <span class="c1"># Show key mutations of BA.1</span>
<span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;test_data/keys/BA2_key.csv&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;BA.2&#39;</span><span class="p">)</span> <span class="c1"># Show key mutations of BA.2</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> <span class="c1"># Show the figure</span>
<span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;modified_profile.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span> <span class="c1"># Save the figure</span>
</code></pre></div>
Below is the result:
<img alt="plot" src="../assets/images/modified_profile.png" /></p>
<p>When key mutations are provided, heatmaps are created below each barplot that marks the mutations (dark gray) belonging to those keys. 
Each row of the heatmap corresponds to the mutations of one key file. A mutation that is present in a key file will always be included in the x-axis even if that mutation is not found in any of the batches of sequences; in that case, an empty column will be shown on the barplot.</p>
<details class="question">
<summary>How does the threshold work?</summary>
<p>The threshold is applied <em>per batch</em>. It sets the minimum count of a mutation in a batch for it to be included. If a mutation meets the threshold in at least one batch, it will be included in the figure as a mutation of that batch but it won't show up in the other batches even if the count is nonzero.</p>
</details>
<details class="question">
<summary>What are key mutations?</summary>
<p>In VARGRAM, key mutations are mutations that will be marked in the mutation profile through a heatmap below the barplots. These key mutations may be co-occuring mutations of a variant or lineage, or just mutations of interest for whatever reason. These are helpful for comparing the batches against a reference set of mutations.</p>
</details>
<p>In addition, you can force the mutation profile to have a horizontal layout. You can also order the genes by providing an annotation file in <a href="https://docs.nextstrain.org/projects/nextclade/en/stable/user/input-files/03-genome-annotation.html">GFF3 format</a>:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="2:2"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><div class="tabbed-labels"><label for="__tabbed_2_1">Sample script</label><label for="__tabbed_2_2">Sample annotation (<code>sc2.gff</code>)</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">vargram</span><span class="w"> </span><span class="kn">import</span> <span class="n">vargram</span> <span class="c1"># Import main module</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">vargram.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">example</span> <span class="c1"># Import module for downloading external data</span>
<span class="n">example</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;test_data&#39;</span><span class="p">)</span> <span class="c1"># Download test data into directory called test_data </span>

<span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;test_data/analysis/omicron_analysis_cli.tsv&#39;</span><span class="p">,</span> <span class="c1"># Provide data</span>
<span class="hll">            <span class="n">gene</span><span class="o">=</span><span class="s1">&#39;test_data/sc2.gff&#39;</span><span class="p">)</span> <span class="c1"># Provide annotation file</span>
</span><span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="c1"># Set minimum count for a mutation to be included</span>
        <span class="n">ytype</span><span class="o">=</span><span class="s1">&#39;counts&#39;</span><span class="p">)</span> <span class="c1"># Set y-axis to show raw count</span>
<span class="n">vg</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">stack_title</span><span class="o">=</span><span class="s1">&#39;Region&#39;</span><span class="p">,</span> <span class="c1"># Change batch legend title</span>
    <span class="n">stack_label</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Foreign&#39;</span><span class="p">,</span> <span class="s1">&#39;Local&#39;</span><span class="p">],</span> <span class="c1"># Change batch names</span>
    <span class="n">stack_color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#009193&#39;</span><span class="p">,</span> <span class="s1">&#39;#E33E84&#39;</span><span class="p">],</span> <span class="c1"># Change batch bar colors</span>
<span class="hll">    <span class="n">order</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="c1"># Order the genes based on the annotation file</span>
</span><span class="hll">    <span class="n">flat</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># Force a horizontal layout</span>
</span><span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;test_data/keys/BA1_key.csv&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;BA.1&#39;</span><span class="p">)</span> <span class="c1"># Show key mutations of BA.1</span>
<span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;test_data/keys/BA2_key.csv&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;BA.2&#39;</span><span class="p">)</span> <span class="c1"># Show key mutations of BA.2</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> <span class="c1"># Show the figure</span>
<span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;ordered_flat_profile.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span> <span class="c1"># Save the figure</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code># Gene map (genome annotation) of SARS-CoV-2 in GFF format.
# For gene map purpses we only need some of the columns. We substitute unused values with &quot;.&quot; as per GFF spec.
# See GFF format reference at https://www.ensembl.org/info/website/upload/gff.html
# seqname   source  feature start   end score   strand  frame   attribute
.   .   gene    26245   26472   .   +   .    gene_name=E
.   .   gene    26523   27191   .   +   .    gene_name=M
.   .   gene    28274   29533   .   +   .    gene_name=N
.   .   gene    266 13468   .   +   .    gene_name=ORF1a
.   .   gene    13468   21555   .   +   .    gene_name=ORF1b
.   .   gene    25393   26220   .   +   .    gene_name=ORF3a
.   .   gene    27202   27387   .   +   .    gene_name=ORF6
.   .   gene    27394   27759   .   +   .    gene_name=ORF7a
.   .   gene    27756   27887   .   +   .    gene_name=ORF7b
.   .   gene    27894   28259   .   +   .    gene_name=ORF8
.   .   gene    28284   28577   .   +   .    gene_name=ORF9b
.   .   gene    21563   25384   .   +   .    gene_name=S
</code></pre></div>
</div>
</div>
</div>
<p>Below is the result:
<img alt="plot" src="../assets/images/ordered_flat_profile.png" /></p>
<p>The ordering of the gene barplots are based on the <code>start</code> column in the annotation file.</p>
</div>
</div>
</div>
<h2 id="basic-syntax">Basic syntax</h2>
<p>VARGRAM has a declarative syntax that makes it is easy to declare what you want to see in the figure. Every VARGRAM script starts by importing the class <code>vargram</code> from the package:
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">vargram</span><span class="w"> </span><span class="kn">import</span> <span class="n">vargram</span> <span class="c1"># Importing vargram class</span>
</code></pre></div></p>
<details class="question">
<summary>What is a class?</summary>
<p>A class is a template for creating an object. To use an analogy, a class is like a blueprint for a car (the object) that has attributes like color and functions (called methods) like driving or windshield wiping. An instance of a class is a built object like a Sedan or an SUV.</p>
<p>The <code>vargram</code> class contains all the methods needed to generate the VARGRAM figure and corresponding summary data. Typically, you would get an instance of the class by defining a variable like <code>vg</code> in the example below. Methods can be accessed by appending <code>.method_name()</code> to the variable, e.g. <code>vg.profile()</code>.</p>
</details>
<p>A mutation profile can then be generated in just three lines of code:
<div class="highlight"><pre><span></span><code><span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;samples-directory&gt;&#39;</span><span class="p">,</span> <span class="c1"># Provide sample sequences</span>
            <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;reference.fa&gt;&#39;</span><span class="p">,</span> <span class="c1"># Provide reference sequence</span>
            <span class="n">gene</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;annotation.gff&gt;&#39;</span><span class="p">)</span> <span class="c1"># Provide genome annotation</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span> <span class="c1"># Tell VARGRAM you want to create a mutation profile</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> <span class="c1"># And show the resulting figure</span>
</code></pre></div>
Here, the path to the sample sequences can be a folder of FASTA files or a single FASTA file. If you already have a Nextclade analysis file, you can provide the path to that file instead:
<div class="highlight"><pre><span></span><code><span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;analysis.csv&gt;&#39;</span><span class="p">),</span> <span class="c1"># Provide Nextclade analysis file (CSV or TSV)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span> <span class="c1"># Tell VARGRAM you want to create a mutation profile</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> <span class="c1"># Show the resulting figure</span>
</code></pre></div>
You can get the dataframe containing the mutations and their counts with:
<div class="highlight"><pre><span></span><code><span class="n">vg</span><span class="o">.</span><span class="n">stat</span><span class="p">()</span>
</code></pre></div>
You can also save the figure by adding:
<div class="highlight"><pre><span></span><code><span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s1">&#39;&lt;my_vargram_figure&gt;.jpg&#39;</span><span class="p">)</span>
</code></pre></div>
You can also save the figure as a PDF or PNG file. Just replace the file extension with what you want.
Finally, you can save the summary data:
<div class="highlight"><pre><span></span><code><span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s1">&#39;&lt;my_vargram_data&gt;.csv&#39;</span><span class="p">)</span>
</code></pre></div></p>
<div class="admonition info">
<p class="admonition-title">Switching the order of <code>show()</code>, <code>save()</code>, and <code>stat()</code></p>
<p>VARGRAM is built on <a href="https://matplotlib.org/">Matplotlib</a> and you may see some similarities in terms of syntax. But you may interchange <code>show()</code> and <code>save()</code> with no worries. In fact, you can save the figure without calling <code>show()</code> or save the summary data without calling <code>stat()</code>. They are all independent of each other but must be called after <code>profile()</code> (see section below about VARGRAM methods).</p>
</div>
<h2 id="vargram-methods">VARGRAM methods</h2>
<h3 id="types">Types</h3>
<p>Broadly, the <code>vargram</code> class has three main types of methods: (1) "plot", (2) "aesthetic" and (3) "terminal" methods. The plot methods determine what type of figure will be generated. At the moment, there's only one plot method, <code>profile()</code>, for generating mutation profiles. </p>
<p>Aesthetic methods like <code>aes()</code> can be used to modify figure characteristics such as labels and colors but other aesthetic methods like <code>key()</code> can also be used to add data aside from the sequence files.</p>
<p>Finally, terminal methods like <code>show()</code>, <code>stat()</code>, and <code>save()</code> produce an output like a generated image, a dataframe, or saved file.</p>
<div class="admonition info">
<p class="admonition-title">Optional methods</p>
<p>All aesthetic methods are optional. However, a plot method and at least one terminal method are required to be called in order to produce an output. A <code>vargram</code> instance must always be defined in order to call these methods.</p>
</div>
<h3 id="order">Order</h3>
<p>The methods should be called in order of their types and form a <strong>PAT</strong> or <strong>PT</strong> sandwich. <strong><em>P</em></strong>lot methods go first, followed by <strong><em>A</em></strong>esthetic methods, and the <strong><em>T</em></strong>erminal methods are called last:
<div class="highlight"><pre><span></span><code><span class="c1"># An instance is defined prior</span>
<span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;analysis.csv&gt;&#39;</span><span class="p">)</span>

<span class="c1"># *** one plot method ***</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
<span class="c1"># *** aesthetic methods ***</span>
<span class="n">vg</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;blue&#39;</span><span class="p">])</span> 
<span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;&lt;my_key&gt;.csv&#39;</span><span class="p">)</span>
<span class="c1"># *** terminal methods ***</span>
<span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s1">&#39;&lt;my_vargram_figure&gt;.pdf&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
Within each method type, the order doesn't matter. Thus, <code>vg.key()</code> can go first before <code>vg.aes()</code> or <code>vg.show()</code> can be called first before <code>vg.save()</code>.</p>
<p>The following examples will <em>not</em> produce an output:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="3:3"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><input id="__tabbed_3_3" name="__tabbed_3" type="radio" /><div class="tabbed-labels"><label for="__tabbed_3_1">A TPA sandwich</label><label for="__tabbed_3_2">An ATP sandwich</label><label for="__tabbed_3_3">A TAP sandwich</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;analysis.csv&gt;&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">stack_title</span><span class="o">=</span><span class="s1">&#39;Region&#39;</span><span class="p">)</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;analysis.csv&gt;&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;&lt;my_key&gt;.csv&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;analysis.csv&gt;&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s1">&#39;&lt;my_vargram_figure&gt;.png&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;&lt;my_key&gt;.csv&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
</code></pre></div>
</div>
</div>
</div>
<p>The following examples <em>will</em> produce an output but the indicated methods will be ignored because they are not inside the PAT or PT sandwich:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="4:2"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><input id="__tabbed_4_2" name="__tabbed_4" type="radio" /><div class="tabbed-labels"><label for="__tabbed_4_1">Before the sandwich</label><label for="__tabbed_4_2">After the sandwich</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;analysis.csv&gt;&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s1">&#39;&lt;my_vargram_data&gt;.csv&#39;</span><span class="p">)</span> <span class="c1"># Ignored</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;analysis.csv&gt;&#39;</span><span class="p">)</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;&lt;my_key&gt;.csv&#39;</span><span class="p">)</span> <span class="c1"># Ignored</span>
</code></pre></div>
</div>
</div>
</div>
<p>You <em>can</em> make more than one PAT sandwich:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="5:2"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><input id="__tabbed_5_2" name="__tabbed_5" type="radio" /><div class="tabbed-labels"><label for="__tabbed_5_1">Same instance</label><label for="__tabbed_5_2">Different instances</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">vg</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;analysis.csv&gt;&#39;</span><span class="p">)</span>

<span class="c1"># PAT sandwich no. 1</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">stack_color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="s1">&#39;blue&#39;</span><span class="p">])</span> 
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;&lt;my_key&gt;.csv&#39;</span><span class="p">)</span> <span class="c1"># Ignored, not inside a PAT sandwich</span>

<span class="c1"># PAT sandwich no. 2</span>
<span class="n">vg</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">stack_color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;orange&#39;</span><span class="p">,</span> <span class="s1">&#39;blue&#39;</span><span class="p">])</span> 
<span class="n">vg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">vg</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">vg1</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;first_analysis.csv&gt;&#39;</span><span class="p">)</span>
<span class="c1"># PAT sandwich no. 1</span>
<span class="n">vg1</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
<span class="n">vg1</span><span class="o">.</span><span class="n">aes</span><span class="p">(</span><span class="n">stack_title</span><span class="o">=</span><span class="s1">&#39;Collecting Institution&#39;</span><span class="p">)</span>
<span class="n">vg1</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="n">vg2</span> <span class="o">=</span> <span class="n">vargram</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s1">&#39;path/to/&lt;second_analysis.csv&gt;&#39;</span><span class="p">)</span>
<span class="n">vg2</span><span class="o">.</span><span class="n">profile</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span> <span class="c1"># Ignored, only one plot method per sandwich--the latest plot method called before the first terminal method</span>
<span class="c1"># PAT sandwich no. 2</span>
<span class="n">vg2</span><span class="o">.</span><span class="n">profile</span><span class="p">()</span>
<span class="n">vg2</span><span class="o">.</span><span class="n">key</span><span class="p">(</span><span class="s1">&#39;&lt;my_key&gt;.csv&#39;</span><span class="p">)</span>
<span class="n">vg2</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
</div>
</div>
</div>
<details class="question">
<summary>Why is it like this?</summary>
<p>Internally, the plot and aesthetic methods only capture their respective inputs and do nothing else. The terminal methods capture their respective inputs, trigger VARGRAM to process the data, and produce their respective outputs. </p>
<p>The first terminal method called in the sandwich notes the latest plot method before it and all other aesthetic methods (if there are any) inside the sandwich. Then, the data is processed accordingly. This is why all other methods outside the sandwich are ignored. Once the data is processed, all the respective outputs (e.g. generated figure, saved data, etc.) of the terminal methods in the sandwich are produced.</p>
</details>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href=".." class="md-footer__link md-footer__link--prev" aria-label="Previous: VARGRAM">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                VARGRAM
              </div>
            </div>
          </a>
        
        
          
          <a href="../mutation_profile/" class="md-footer__link md-footer__link--next" aria-label="Next: Generating a mutation profile">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Generating a mutation profile
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; 2024-2025 <a href="https://pgc.up.edu.ph"  target="_blank" rel="noopener">Philippine Genome Center</a>

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.footer", "content.code.copy", "content.code.annotate"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>